Add woocommerce-install site-picker step

lsl opened the pull request:
Re-introduces the site-picker removed in #58561

Todo:
- Loading /confirm without a site slug should redirect to the site-picker - unless you're coming from the site-picker which won't have the site query param (this can be done but requires a secondary dependency set by the site-picker that isn't modified by the controller like `site` is)
- Hide the confirm step while redirecting to the site-picker
- Hide p2 sites p1638330066305700-slack-C02DQP0FP selectors this is using is not enough to block p2 sites - they're free plans, need to also check they're not wpforteams

#### Changes proposed in this Pull Request

* Replaces siteSlug/siteId with new `site` query param
* Moves confirm step to step 0 so it becomes the default
* If site isn't known, it redirects to the select-site step to pick a site
* Backurl clean up
* Nav button hiding on select step (no way to know where back is)
* Hide step count as it no longer makes sense (1, 2, done of 3)

#### Testing instructions

* `http://calypso.localhost:3000/start/woocommerce-install/select-site` should let you select a site. Back should be disabled.
* `http://calypso.localhost:3000/start/woocommerce-install/confirm?site=example.com` should let you skip the select-site step. Back should take you to `/woocommerce-installation/example.com`
* `http://calypso.localhost:3000/start/woocommerce-install/confirm` should redirect to `/select-site` but this will only happen on a fresh state, hitting back button here should load `/select-site`.
* `http://calypso.localhost:3000/start/woocommerce-install/select-site` at this point should load the picker and clear the stored site, now going to `/confirm` will redirect back to `/select-site`

Related to #58229

lsl commented on client/signup/steps/woocommerce-install/confirm/index.tsx:
```diff
@@ -129,7 +139,14 @@ export default function Confirm( props: WooCommerceInstallProps ): ReactElement
-					<NextButton onClick={ () => goToStep( 'transfer' ) }>{ __( 'Confirm' ) }</NextButton>
+					<NextButton
+						onClick={ () => {
+							dispatch( submitSignupStep( { stepName: 'confirm' }, { confirm: true } ) );
```
Want to prevent people landing on transfer without a confirm step happening. (they can bypass with siteslug param)

retrofox replied:
i
```suggestion
							dispatch( submitSignupStep( { stepName: 'confirm' }, { confirmed: true } ) );
```

lsl replied:
thanks, renamed this further in the end.

lsl commented on client/signup/steps/woocommerce-install/select-site/index.tsx:
```diff
@@ -45,15 +44,18 @@ export default function SelectSite( props: WooCommerceInstallProps ): ReactEleme
+		dispatch( submitSignupStep( { stepName: 'select-site' }, { siteSlug } ) );
 		goToStep( 'confirm' );
-	}, [ selectedSiteId, goToStep, dispatch ] );
+	}, [ siteSlug, goToStep, dispatch ] );
```
Was attempting to reduce confusion around siteId/slug but this didn't seem to matter too much in the end.

p-jackson replied:
It matters for the Hero Flow but don't think it does for `woocommerce-install` 🤔 
We might create a Hero Flow url using an old site slug, then after going through checkout the site slug changes (because of a domain purchase), which invalidates the site slug. That's why sometimes we care about site ID.
But I think for woop you're never coming straight from a domain purchase right?

lsl replied:
Maybe one day but I doubt it at this point (steps aren't really designed to be interchangeable like I expected they would be.)

lsl commented on client/signup/controller.js:
```diff
@@ -289,14 +287,6 @@ export default {
-			! providesDependenciesInQuery?.includes( 'siteSlug' )
-		) {
-			context.store.dispatch( setSelectedSiteId( null ) );
-		}
```
I'm not clear why this was being set to null, which situations did this need to happen? Domains? 

Instead I'm suggesting moving this clear to only happen when there is no site id or slug in the query or dependencies passed in.

I think that's same result but its more clear why/when its happening? (maybe :thinking: )

retrofox replied:
Is it a generic concern? Because I think it won't affect the woocommerce-install flow since `providesDependenciesInQuery` includes one of them.

p-jackson replied:
This call to `setSelectedSiteId( null )` was first introduced in #42162, which was something to do with getting the pricing correct.

I don't think changing this to take into account both the query _and_ dependencies is equivalent to what we're doing right now. If you look at #42162 the testing instructions talk about using a URL that includes a site slug, but it's not in the query param. So this change you're proposing could end up clearing a selected site ID that was passed on the URL, just not in a query param.

I don't know what the implications of this are, I just am pointing out that it could be a change in behaviour and not necessarily a safe refactor. So if it doesn't affect the `woocommerce-install` flow then I'm not sure about changing it.

lsl replied:
My reasoning for removing it was that it doesn't make sense - these could be optional, if they're not provided it should be clearing the selected site no? Maybe it doesn't matter in practice?

lsl commented on client/signup/controller.js:
```diff
@@ -318,11 +308,13 @@ export default {
-			query?.siteId;
+			query?.siteId ||
+			signupDependencies?.siteSlug ||
+			signupDependencies?.siteId;
```
The order here was a bit weird, if a user puts a siteid or slug in the URL we should respect and use that - always. Am I misunderstanding where this is coming from?

p-jackson replied:
I think your instinct here to prefer the query param over the redux store makes sense. But I'd just want to check that this doesn't break our "choose domain" -> "checkout" -> "hero flow" progression, which is why we added this code in the first place.

lsl commented on client/signup/steps/woocommerce-install/confirm/index.tsx:
```diff
@@ -16,16 +16,26 @@ import {
+		isReskinned,
+		stepSectionName,
+		headerTitle,
+		headerDescription /*, siteId*/,
```
I don't know how much I trust siteId coming from props, to me it looks like it could get changed in the controller async (.then()...). I might be wrong on this, couldn't quite figure it out. If I'm right the transfer step is risky until we know the site Id is certain and not going to change.

p-jackson replied:
The reason we're fetching the site again is in case the site slug changed due to a domain being purchased. So I suspect if any prop changed it'd be the `siteSlug`, not `siteId`

lsl replied:
Figured out what I was misunderstanding here, was non issue, its not calling next until the selected site is dispatched again.

Also site id isn't passed through props directly but `queryObject` or `signupDependencies`.

lsl committed:
Add site picker step to woocommerce-install

lsl commented:
@p-jackson / @retrofox reworked this to completely side step selectedSiteId/siteId/siteSlug as it felt a lot less brittle. Instead I set up a new `site` dependency that doesn't rely on magic happening in the controller. This is set by the site picker or the query param, the query param overrides the dependency value in the confirm step, the confirm step "confirms" a site id and passes that off to the transfer step. I did this final bit as a to prevent anyone thinking its a good idea to pass the same dependency all the way through or trying to preconfirm a transfer using the query param.

All this might still be for naught, I don't see an easy way to make a step optional if dependencies are met via query params. The process will redirect to the site picker step if a site hasn't been selected using the picker yet, you can set one through the picker and override it through the query param for testing but this isn't the final behavior we want.

lsl commented:
Unless there is a way to make steps optional we'll just drop this completely. https://github.com/Automattic/wp-calypso/pull/58561

lsl commented on client/signup/steps/woocommerce-install/select-site/index.tsx:
```diff
@@ -43,17 +43,21 @@ export default function SelectSite( props: WooCommerceInstallProps ): ReactEleme
+	const [ selected, setSelected ] = useState();
 
+	const site = useSelector( ( state ) => getSiteSlug( state, selected ) );
+	const siteId = useSelector( ( state ) => getSiteId( state, site ) as number );
```
:thinking:  this could probably be tidied up

lsl commented:
@p-jackson / @retrofox managed to get this "working":

- No way to enter the signup flow midway through e.g. in step 1 with a siteSlug provided, to work around this I moved confirm to step 0, moved site-select step to step 3. In the confirm step, [in the controller] I check for a site slug in the url, if this is present, it sets the selected site id and the confirm screen loads. If there is no slug in the url, it goes to the select site step which returns you to the confirm step with the selected site id set.
- To work around the [persistence issue](https://github.com/Automattic/wp-calypso/pull/58561#issuecomment-981936762) I had to make the query param take precedence over the persisted value, since we ?can't/won't? swap the order of precedence I instead just added a `site` param that that overrides siteId and siteSlug in the query params.

Still not great:
- Flashing the confirm screen before goTo select-site shouldn't
- Step count in the top right isn't calculated right anymore, this needs hiding or deferring to the step itself to decide
- Unsure how resume will work with the checkout flow we want to add - I think it'll be ok.
- It'd be good if this select site step was usable across the board, I think that's possible by setting a return step as a dependency for the select site step that can then be set to decide where it sends you afterward. Keeping this step in its own flow with no step count might help fix other issues too.

p-jackson commented:
> I don't see an easy way to make a step optional if dependencies are met via query params

There used to be an old flow that did this, but maybe that code doesn't exist anymore. This is how the `onboarding` flow removes the `user` step when they're already logged in.
https://github.com/Automattic/wp-calypso/blob/75c8035911b510a63903e6d60cab449922c528a8/client/signup/config/flows.js#L205

Seems sort of hard coded, so maybe not appropriate for this site selector thingy.

lsl commented:
We could add some kind of callback hook to run before the controller gets its fingers into things. That would let each flow to define its own dynamic steps as required.

lsl commented:
Todo:
- Can we redirect to the landing page if there’s no siteID?
- Hide the confirm screen until fully loaded / redirect happens

pablinos commented on client/signup/config/flows-pure.js:
```diff
@@ -453,12 +453,12 @@ export function generateFlows( {
-			providesDependenciesInQuery: [ 'siteSlug' ],
-			optionalDependenciesInQuery: [ 'siteSlug' ],
-			disallowResume: true,
+			providesDependenciesInQuery: [ 'site' ],
```
It's a shame we can't use `siteSlug` as it seems to be handled in much the same way. Does it work or does the controller stuff break things?

lsl replied:
`signupDependencies.siteSlug` [takes precendence](https://github.com/Automattic/wp-calypso/blob/1ae846fd6d6e09cf7897a2e4aad5ce35bd14cb0e/client/signup/controller.js#L320-L324) over `queryObject.siteSlug` and talking to @p-jackson we weren't comfortable changing this around. Adding `queryObject.site` that takes precedence over `signupDependencies.site` and `siteSlug`/`siteId` seemed like the safest option.

pablinos commented on client/signup/steps/woocommerce-install/select-site/index.tsx:
```diff
@@ -43,31 +42,38 @@ export default function SelectSite( props: WooCommerceInstallProps ): ReactEleme
 			className="select-site__step-wrapper"
 			align={ isReskinned ? 'left' : 'center' }
 			isWideLayout={ isReskinned }
+			headerText={ title }
+			fallbackHeaderText={ title }
+			subHeaderText={ subtitle }
+			fallbackSubHeaderText={ subtitle }
```
I'm not sure, but I feel like these were removed to fix something related to the confirm screen. It all seems to be working at the moment though.

lsl replied:
Suspect similar change needs to happen in the confirm screen, (moving them out of the flow file and into the step itself) but its not currently causing an error in any case could find.

pablinos reviewed (approved):
This tested as expected apart from (maybe):

> http://calypso.localhost:3000/start/woocommerce-install/select-site at this point should load the picker 

I didn't have fresh state in the previous step. I went to `select-site`, then went directly to `confirm`. It loaded it for a previously selected site. Should going to `select-site` always clear the state? If not, then I'd say this is good to go.

lsl commented:
> I didn't have fresh state in the previous step. I went to select-site, then went directly to confirm. It loaded it for a previously selected site. Should going to select-site always clear the state? If not, then I'd say this is good to go.

This isn't ideal, the ~controller~ main [redirects](https://github.com/Automattic/wp-calypso/blob/98691b2a629096284030a4c2888eefad95a8755e/client/signup/main.jsx#L196-L203) to step 0 when you try to load a step that isn't step 0. So you're getting redirected to `/confirm` which then, because of shared magic state, you stay there with the selected site.

This seems like a bug we can live with but I'm worried it's going to complicate things when we want to add extra onboarding steps later.

lsl commented:
Decided to remove the site picker completely in #58561 and then rebase and re-introduce it here.

I don't want to waste anymore time on this so shelving it with some todo's mentioned in the PR.

If we want this later when we start adding more steps and looking at this dependency stuff again, the PR is here, for now, lets move on.

There was also some shared fixes so it sort of makes sense to stage the changes like this anyway.

lsl commented on client/signup/steps/woocommerce-install/select-site/index.tsx:
```diff
@@ -0,0 +1,82 @@
+	const filterSites = ( site: Site ) => {
+		return (
+			site.capabilities.manage_options &&
+			( isFreePlan( site.plan.product_slug ) ||
+				isPersonalPlan( site.plan.product_slug ) ||
+				isPremiumPlan( site.plan.product_slug ) ||
+				isBusinessPlan( site.plan.product_slug ) ||
+				isEcommercePlan( site.plan.product_slug ) )
```
p1638330066305700-slack-C02DQP0FP not enough to block p2 sites - they're free plans, need to also check they're not wpforteams
